// @livestore/sync-cf@0.3.0/dist/common/ws-message-types.js
import { LiveStoreEvent } from '@livestore/common/schema';
import { Schema } from '@livestore/utils/effect';
export const PullReq = Schema.TaggedStruct('WSMessage.PullReq', {
    requestId: Schema.String,
    /** Omitting the cursor will start from the beginning */
    cursor: Schema.optional(Schema.Number),
}).annotations({ title: '@livestore/sync-cf:PullReq' });
export const SyncMetadata = Schema.Struct({
    /** ISO date format */
    createdAt: Schema.String,
}).annotations({ title: '@livestore/sync-cf:SyncMetadata' });
export const PullRes = Schema.TaggedStruct('WSMessage.PullRes', {
    batch: Schema.Array(Schema.Struct({
        eventEncoded: LiveStoreEvent.AnyEncodedGlobal,
        metadata: Schema.Option(SyncMetadata),
    })),
    requestId: Schema.Struct({ context: Schema.Literal('pull', 'push'), requestId: Schema.String }),
    remaining: Schema.Number,
}).annotations({ title: '@livestore/sync-cf:PullRes' });
export const PushReq = Schema.TaggedStruct('WSMessage.PushReq', {
    requestId: Schema.String,
    batch: Schema.Array(LiveStoreEvent.AnyEncodedGlobal),
}).annotations({ title: '@livestore/sync-cf:PushReq' });
export const PushAck = Schema.TaggedStruct('WSMessage.PushAck', {
    requestId: Schema.String,
}).annotations({ title: '@livestore/sync-cf:PushAck' });
export const Error = Schema.TaggedStruct('WSMessage.Error', {
    requestId: Schema.String,
    message: Schema.String,
}).annotations({ title: '@livestore/sync-cf:Error' });
export const Ping = Schema.TaggedStruct('WSMessage.Ping', {
    requestId: Schema.Literal('ping'),
}).annotations({ title: '@livestore/sync-cf:Ping' });
export const Pong = Schema.TaggedStruct('WSMessage.Pong', {
    requestId: Schema.Literal('ping'),
}).annotations({ title: '@livestore/sync-cf:Pong' });
export const AdminResetRoomReq = Schema.TaggedStruct('WSMessage.AdminResetRoomReq', {
    requestId: Schema.String,
    adminSecret: Schema.String,
}).annotations({ title: '@livestore/sync-cf:AdminResetRoomReq' });
export const AdminResetRoomRes = Schema.TaggedStruct('WSMessage.AdminResetRoomRes', {
    requestId: Schema.String,
}).annotations({ title: '@livestore/sync-cf:AdminResetRoomRes' });
export const AdminInfoReq = Schema.TaggedStruct('WSMessage.AdminInfoReq', {
    requestId: Schema.String,
    adminSecret: Schema.String,
}).annotations({ title: '@livestore/sync-cf:AdminInfoReq' });
export const AdminInfoRes = Schema.TaggedStruct('WSMessage.AdminInfoRes', {
    requestId: Schema.String,
    info: Schema.Struct({
        durableObjectId: Schema.String,
    }),
}).annotations({ title: '@livestore/sync-cf:AdminInfoRes' });
export const Message = Schema.Union(PullReq, PullRes, PushReq, PushAck, Error, Ping, Pong, AdminResetRoomReq, AdminResetRoomRes, AdminInfoReq, AdminInfoRes).annotations({ title: '@livestore/sync-cf:Message' });
export const BackendToClientMessage = Schema.Union(PullRes, PushAck, AdminResetRoomRes, AdminInfoRes, Error, Pong);
export const ClientToBackendMessage = Schema.Union(PullReq, PushReq, AdminResetRoomReq, AdminInfoReq, Ping);
