// @signaldb/core@1.6.0/dist/index.cjs.js
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const Cursor = require("./index.cjs2.js");
const index = require("./index.cjs3.js");
const AutoFetchCollection = require("./index.cjs4.js");
const combinePersistenceAdapters = require("./index.cjs5.js");
const createIndexProvider = require("./index.cjs6.js");
const createMemoryAdapter = require("./index.cjs7.js");
const createPersistenceAdapter = require("./index.cjs8.js");
const createReactivityAdapter = require("./index.cjs9.js");
const isEqual = require("./index.cjs10.js");
const modify = require("./index.cjs11.js");
const randomId = require("./index.cjs12.js");
const EventEmitter = require("./index.cjs13.js");
const createIndex = require("./index.cjs14.js");
exports.Cursor = Cursor.default;
exports.Collection = index.default;
exports.AutoFetchCollection = AutoFetchCollection;
exports.combinePersistenceAdapters = combinePersistenceAdapters.default;
exports.createIndexProvider = createIndexProvider;
exports.createMemoryAdapter = createMemoryAdapter;
exports.createPersistenceAdapter = createPersistenceAdapter;
exports.createReactivityAdapter = createReactivityAdapter;
exports.isEqual = isEqual;
exports.modify = modify;
exports.randomId = randomId;
exports.EventEmitter = EventEmitter;
exports.createIndex = createIndex.default;
